### ISSUE TITLE
Bluetooth support for Ubuntu 21.10 ?

### ISSUE THREAD

**Issue by fetushito** [t2linux/wiki#67]
The guide doesn't mention anything about Bluetooth. Is Bluetooth somehow assumed to work out of the box? I have seen the driver works perfectly on Ubuntu 20.04 on my MBP 16,1, but after migrating to 21.10 it no longer works. Currently running kernel 5.11.10 with @aunali1 patches.

The 'demsg commad |grep Bluetooth' shows the following:

[   24.657447] Bluetooth: hci0: BCM: failed to write update baudrate (-16)
[   24.657460] Bluetooth: hci0: Failed to set baudrate
[   24.780703] Bluetooth: hci0: BCM: chip id 150
[   24.781514] Bluetooth: hci0: BCM: features 0x07
[   24.783055] Bluetooth: hci0: BCM4364B3 Trinidad Olympic GEN (MFG)
[   24.783060] Bluetooth: hci0: BCM (001.016.065) build 0092
[   24.783092] Bluetooth: hci0: BCM: firmware Patch file not found, tried:
[   24.783095] Bluetooth: hci0: BCM: 'brcm/BCM.hcd'

Any suggestion are greatly appreciated.

**Comment by Redecorating** [comment]
bluetooth is broken by 5.11, you can use the 5.10 16x-wifi kernel in this build artifact which will have wifi and bluetooth both working <https://github.com/marcosfad/mbp-ubuntu-kernel/actions/runs/1014749493>

**Comment by fetushito** [comment]
Thanks @Redecorating. I've never installed a kernel through the .deb files. Is there any way I can use them to do it the classic way? (Downloading vanilla kernel from kernel.org, applying the patches and building it)

**Comment by Redecorating** [comment]
If you're doing that, then just download the kernel source for 5.10, and use patches from this commit https://github.com/jamlam/mbp-16.1-linux-wifi/tree/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4
(this repo has support for wifi on the 16,1)

**Comment by fetushito** [comment]
Thanks again. Tried that one but hit another issue I had previously seen when trying to compile other versions:

drivers/hwmon/applesmc.c: In function ‘applesmc_init_smcreg_try’:
drivers/hwmon/applesmc.c:829:34: error: ‘smcreg’ undeclared (first use in this function)
  829 |  struct applesmc_registers *s = &smcreg;
      |                                  ^~~~~~
drivers/hwmon/applesmc.c:829:34: note: each undeclared identifier is reported only once for each function it appears in
drivers/hwmon/applesmc.c:838:28: error: ‘smc’ undeclared (first use in this function)
  838 |  ret = read_register_count(smc, &count);

**Comment by Redecorating** [comment]
That's the same as this issue https://github.com/jamlam/mbp-16.1-linux-wifi/issues/1. Have all the applesmc patches applied correctly? What flags did you use for `patch`?

Line 829 in drivers/hwmon/applesmc.c is `struct applesmc_registers *s = &smc->reg;` when I apply the patches with `makepkg`, which applies them with `patch -Np1 < "$current-patch"`, and it does the patches in the order that they are numbered.

**Comment by fetushito** [comment]
This is how I apply the patch:
`for i in /home/[USER]/wifi_drivers_kernel/mbp-16.1-linux-wifi/*.patch; do patch -p1 --verbose < $i; done`

LIne 829 is no getting changed. Is that an indication that the patch is wrong or I don't have the right kernel version? (base kernel that I am using: https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.10.tar.xz)

I managed to locate the patch rejection, this is how a portion of it looks:

`--- drivers/hwmon/applesmc.c
+++ drivers/hwmon/applesmc.c
@@ -578,9 +594,9 @@ static int applesmc_init_index(struct applesmc_registers *s)
 /*
  * applesmc_init_smcreg_try - Try to initialize register cache. Idempotent.
  */
-static int applesmc_init_smcreg_try(void)
+static int applesmc_init_smcreg_try(struct applesmc_device *smc)
 {
-       struct applesmc_registers *s = &smcreg;
+       struct applesmc_registers *s = &smc->reg;
        bool left_light_sensor = false, right_light_sensor = false;
        unsigned int count;
        u8 tmp[1];
`

Not sure if there's a way to adapt the patches, that's a bit out of my area of expertise.

**Comment by AdityaGarg8** [comment]
Jamlam's repo has been updated for 5.13
The patches don’t work for 5.10 now

**Comment by AdityaGarg8** [comment]
You can install using deb file by typing `sudo apt install ` and then drag and drop the linux headers deb to the terminal and press enter

**Comment by AdityaGarg8** [comment]
Then do the same for linux images

**Comment by AdityaGarg8** [comment]
and lastly the same for linux libc dev

**Comment by AdityaGarg8** [comment]
Link of kernel https://github.com/Redecorating/mbp-ubuntu-kernel/releases/tag/v5.10.47-1

**Comment by AdityaGarg8** [comment]
Also, 20.10 uses 5.11 I guess, so GRUB may boot that instead. So you may either remove it or use advanced options in GRUB

**Comment by fetushito** [comment]
> Jamlam's repo has been updated for 5.13
> The patches don’t work for 5.10 now

So using 5.13 as a base kernel would do it? I kinda don't like the idea of just copy pasting .deb files, as I don't know whether there are security implications. (Mainly because I don't know how to inspect the package's content)

**Comment by AdityaGarg8** [comment]
Bluetooth won't work on 5.13. I also have a 16,1 and 5.10 was the last kernel in which Bluetooth worked for me

**Comment by AdityaGarg8** [comment]
The build.sh script in this [link ](https://github.com/Redecorating/mbp-ubuntu-kernel) is used to compile the kernel in deb files

**Comment by AdityaGarg8** [comment]
You may read the script

**Comment by AdityaGarg8** [comment]
[https://github.com/jamlam/mbp-16.1-linux-wifi/tree/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4](https://github.com/jamlam/mbp-16.1-linux-wifi/tree/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4)

That's jamlam's repo's older version for 5.10

**Comment by AdityaGarg8** [comment]
Redecorating used this to compile 5.10.47

**Comment by AdityaGarg8** [comment]
I guess you won't face issues with this

**Comment by fetushito** [comment]
After installing the .deb files initially suggested, bluetooth doesn't work nor does the wifi. Will try with 5.10 and the patches you indicate (https://github.com/jamlam/mbp-16.1-linux-wifi/tree/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4)

**Comment by Redecorating** [comment]
Have you rebooted with the new kernel? You should have `uname -r` saying something like `5.10.47-mbp-16x-wifi`.

Once you are running the patched kernel, WiFi won't work until you install wifi firmware for the driver to use: https://wiki.t2linux.org/guides/wifi/

If you still want to build and patch it from source, this patches it correctly on ubuntu for me, were you doing anything particularly different from this? (it also worked with 5.10 kernel source instead of 5.10.50)
```
git clone https://github.com/jamlam/mbp-16.1-linux-wifi
cd mbp-16.1-linux-wifi/
git checkout c0d4
wget https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.10.50.tar.xz
tar -xf linux-5.10.50.tar.xz
cd linux-5.10.50/
for i in ../*.patch; do patch -p1 --verbose < $i;done
```
This resulted in that section being patched correctly.
```c
/*
 * applesmc_init_smcreg_try - Try to initialize register cache. Idempotent.
 */
static int applesmc_init_smcreg_try(struct applesmc_device *smc)
{
	struct applesmc_registers *s = &smc->reg;
	bool left_light_sensor = 0, right_light_sensor = 0;
```

**Comment by fetushito** [comment]
No, nothing different than that. The winner combo at the end was what @AdityaGarg8 advised:
Kernel: https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.10.47.tar.xz
Patches (originally for 5.10.33 I think): https://github.com/jamlam/mbp-16.1-linux-wifi/tree/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4

Plus a little bit of magic from https://wiki.t2linux.org/guides/dkms/

I want to thank you both @Redecorating and @AdityaGarg8 for your help. I finally have this thing working!!!!

**Comment by AdityaGarg8** [comment]
Most welcome ☺️

**Comment by AdityaGarg8** [comment]
BTW can you share your way to compile kernels

**Comment by AdityaGarg8** [comment]
I mean the procedure with codee

**Comment by fetushito** [comment]
The procedure looks something like this:

1.- `mkdir workingdir`
2.- `cd workingdir`
3.-  `wget https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.10.47.tar.xz`
4.- `tar xvf linux-5.10.47.tar.xz`
5.- `wget https://github.com/jamlam/mbp-16.1-linux-wifi/archive/c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4.zip`
6.- `unzip mbp-16.1-linux-wifi-c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4.zip`
7.- `cd linux-5.10.47`
8.- `cp /boot/config-5.10.47 ./.config`
9.- Inside .config file, change `CONFIG_SYSTEM_TRUSTED_KEYS="debian/canonical-certs.pem` to `CONFIG_SYSTEM_TRUSTED_KEYS=""`
10.- `for i in /home/$USER/Desktop/workingdir/mbp-16.1-linux-wifi-c0d492b15c227d3dfc06fbcad3335e1a5aa00ca4/*.patch ; do patch -p1 --verbose < $i; done`
11.- `make -j 16`
12.- `make INSTALL_MOD_STRIP=1 modules_install -j 16`
13.- `make install -j 16`
14.- `update initramfs -u`

Let me know if you have any feedback on how to make it better.

**Comment by AdityaGarg8** [comment]
Thanks

**Comment by AdityaGarg8** [comment]
I shall try this out and give you the feedback

**Comment by AdityaGarg8** [comment]
Can you please list the packages you are using to compile. Like gcc for a start

**Comment by AdityaGarg8** [comment]
Instead of `cp /boot/config-5.10.47 ./.config` you can use `make menuconfig`

**Comment by AdityaGarg8** [comment]
And your code worked for 5.10.50 too!

But do share the packages you use to build kernels

**Comment by fetushito** [comment]
These are the ones I recall (followed [these](https://gist.github.com/gbrow004/096f845c8fe8d03ef9009fbb87b781a4#compiling-a-new-kernel-with-patches-to-complete-almost-the-functionality-of-your-mbp) instructions):

bison/hirsute,now 2:3.7.5+dfsg-1 amd64 [installed]
flex/hirsute,now 2.6.4-8 amd64 [installed]
gcc/hirsute,now 4:10.3.0-1ubuntu1 amd64 [installed]
libncurses-dev/hirsute,now 6.2+20201114-2build1 amd64 [installed]
libssl-dev/hirsute,now 1.1.1j-1ubuntu3 amd64 [installed]
make/hirsute,now 4.3-4ubuntu1 amd64 [installed]

**Comment by fetushito** [comment]
> Instead of `cp /boot/config-5.10.47 ./.config` you can use `make menuconfig`

What is the difference here? I thought using a config file from a previous ubuntu-specific kernel build would make more sense compared to using default options.

**Comment by AdityaGarg8** [comment]
I didn’t know that you wanted the one for ubuntu specific kernel
